### ISSUE TITLE
[YSQL] Attempted to restart when session does not require restart

### ISSUE THREAD

**Issue by pkj415** [yugabyte/yugabyte-db#20735]
Jira Link: [DB-9738](https://yugabyte.atlassian.net/browse/DB-9738)
### Description

Repro example: (to be added from slack)

RCA:
`Attempted to restart when session does not require restart` occurs when the following events occur:
(1) Pg sends a read rpc X to PgClientSession
(2) Pg sends another read rpc Y to PgClientSession without waiting for X
(3) X results in a read restart error from the tserver, so it ends up updating `restarts_` in the ConsistentReadPoint used for the session.
(4) Y arrives at the PgClientSession and updates the read time which reaches SetReadTimeUnlocked() (resulting in `restarts_` being cleared).
(5) X's response reaches Pg and it results in a retry for the whole statement with the new restart read point. But it fails at the SCHECK because `restarts_` is empty.

Orthogonally, I also see the following warnings in the Pg logs:
```
W0122 16:39:48.741339 280802 pg_session.cc:173] Operations in a batch have different read times { read: { physical: 1705941588715739 logical: 4095 } local_limit: { physical: 1705941588715739 logical: 4095 } global_limit: { physical: 1705941588784210 } in_txn_limit: { physical: 1705941588732530 logical: 6 } serial_no: 0 } vs { read: { physical: 1705941588715739 logical: 4095 } local_limit: { physical: 1705941588784210 } global_limit: { physical: 1705941588784210 } in_txn_limit: { physical: 1705941588732530 logical: 6 } serial_no: 0 }
```

### Issue Type

kind/bug

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-9738]: https://yugabyte.atlassian.net/browse/DB-9738?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by tsetiawan-yugabyte** [comment]
how to reproduce:
1. continuously update records on a table
2. execute select query that have a complex join:

  SELECT "row_number"() OVER(ORDER BY "testDescription"),
                    "testDescription"
                FROM "testMaster" master
                LEFT JOIN "testJoin" tj ON tj."testJoinKey" = master."testLink"
        LEFT JOIN "testLevel" lvl ON lvl."testNumber" = tj."testJoinLevelLink"
		LEFT JOIN "testLevel" lvl2 ON lvl2."testNumber" = lvl."testNextNumber"
                WHERE lvl2."testNextNumber" like '%A'
                LIMIT 1 ;\watch 1

**Comment by pao214** [comment]
Possibly solved by #22821. Please re-open if the issue occurs again.
